cmd/config-current-notify.go

```diff
@@ -0,0 +1,924 @@
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package cmd
+
+import (
+	"context"
+	"crypto/tls"
+	"fmt"
+	"strconv"
+	"strings"
+	"time"
+
+	"github.com/minio/minio/cmd/logger"
+	"github.com/minio/minio/pkg/event/target"
+	xnet "github.com/minio/minio/pkg/net"
+)
+
+// Notification config constants.
+const (
+	notifyKafkaSubSys    = "notify.kafka"
+	notifyMQTTSubSys     = "notify.mqtt"
+	notifyMySQLSubSys    = "notify.mysql"
+	notifyNATSSubSys     = "notify.nats"
+	notifyNSQSubSys      = "notify.nsq"
+	notifyESSubSys       = "notify.elasticsearch"
+	notifyAMQPSubSys     = "notify.amqp"
+	notifyPostgresSubSys = "notify.postgres"
+	notifyRedisSubSys    = "notify.redis"
+	notifyWebhookSubSys  = "notify.webhook"
+
+	// Add new constants here if you add new fields to config.
+)
+
+func (s serverConfig) SetNotifyKafka(kName string, cfg target.KafkaArgs) error {
+	if err := cfg.Validate(); err != nil {
+		return err
+	}
+
+	s[notifyKafkaSubSys][kName] = []KV{
+		{
+			Key: stateKey,
+			Value: func() string {
+				if cfg.Enable {
+					return stateEnabled
+				}
+				return stateDisabled
+			}(),
+		},
+		{
+			Key: "brokers",
+			Value: func() string {
+				var brokers []string
+				for _, broker := range cfg.Brokers {
+					brokers = append(brokers, broker.String())
+				}
+				return strings.Join(brokers, ",")
+			}(),
+		},
+		{
+			Key:   "topic",
+			Value: cfg.Topic,
+		},
+		{
+			Key:   "queueDir",
+			Value: cfg.QueueDir,
+		},
+		{
+			Key:   "queueLimit",
+			Value: strconv.Itoa(int(cfg.QueueLimit)),
+		},
+		{
+			Key:   "tls.enable",
+			Value: strconv.FormatBool(cfg.TLS.Enable),
+		},
+		{
+			Key:   "tls.skipVerify",
+			Value: strconv.FormatBool(cfg.TLS.SkipVerify),
+		},
+		{
+			Key:   "tls.clientAuth",
+			Value: strconv.Itoa(int(cfg.TLS.ClientAuth)),
+		},
+		{
+			Key:   "sasl.enable",
+			Value: strconv.FormatBool(cfg.SASL.Enable),
+		},
+		{
+			Key:   "sasl.user",
+			Value: cfg.SASL.User,
+		},
+		{
+			Key:   "sasl.password",
+			Value: cfg.SASL.Password,
+		},
+	}
+	return nil
+}
+
+func (s serverConfig) GetNotifyKafka() map[string]target.KafkaArgs {
+	kafkaTargets := make(map[string]target.KafkaArgs)
+	for k, kvs := range s[notifyKafkaSubSys] {
+		enabled := kvs.Get(stateKey) == stateEnabled
+		if enabled {
+			var err error
+			var brokers []xnet.Host
+			for _, s := range strings.Split(kvs.Get("brokers"), ",") {
+				var host *xnet.Host
+				host, err = xnet.ParseHost(s)
```

Comment:
It is better to trim space left and right in `s` first (`s = strings.TrimSpace(s)`)